src/vulkan-renderer/camera.cpp

```diff
@@ -1,160 +1,190 @@
+               const float window_height)
+    : m_position(position), m_yaw(yaw), m_pitch(pitch) {
+    set_aspect_ratio(window_width, window_height);
+    update_vectors();
+    update_matrices();
 }
 
-void Camera::update_view_matrix() {
-    glm::mat4 rot_m = glm::mat4(1.0f);
-    glm::mat4 trans_m;
+void Camera::update_vectors() {
+    if (m_type == CameraType::CAMERA_TYPE_LOOKAT) {
+        glm::vec3 front;
+        front.x = cos(glm::radians(m_yaw)) * cos(glm::radians(m_pitch));
+        front.y = sin(glm::radians(m_pitch));
+        front.z = sin(glm::radians(m_yaw)) * cos(glm::radians(m_pitch));
+
+        // Normalize the vectors, because their length gets closer to 0 the
+        // more you look up or down which results in slower movement.
+        m_front = glm::normalize(front);
+        m_right = glm::normalize(glm::cross(m_front, m_world_up));
+        m_up = glm::normalize(glm::cross(m_right, m_front));
+        m_update_needed = true;
+    }
+}
 
-    rot_m = glm::rotate(rot_m, glm::radians(m_rotation.x), glm::vec3(1.0f, 0.0f, 0.0f));
-    rot_m = glm::rotate(rot_m, glm::radians(m_rotation.y), glm::vec3(0.0f, 1.0f, 0.0f));
-    rot_m = glm::rotate(rot_m, glm::radians(m_rotation.z), glm::vec3(0.0f, 0.0f, 1.0f));
+void Camera::update_matrices() {
+    if (m_type == CameraType::CAMERA_TYPE_LOOKAT) {
+        m_view_matrix = glm::lookAt(m_position, m_position + m_front, m_up);
+        m_perspective_matrix = glm::perspective(glm::radians(m_fov), m_aspect_ratio, m_near_plane, m_far_plane);
+        m_update_needed = false;
+    }
+}
 
-    trans_m = glm::translate(glm::mat4(1.0f), m_position * glm::vec3(1.0f, 1.0f, -1.0f));
+glm::mat4 Camera::get_view_matrix() {
+    if (m_update_needed) {
+        update_matrices();
+    }
+    return m_view_matrix;
+}
 
-    if (m_type == CameraType::FIRSTPERSON) {
-        m_matrices.view = rot_m * trans_m;
-    } else {
-        m_matrices.view = trans_m * rot_m;
+glm::mat4 Camera::get_perspective_matrix() {
+    if (m_update_needed) {
+        update_matrices();
     }
+    return m_perspective_matrix;
+}
 
-    m_updated = true;
-};
+bool Camera::is_moving() const {
+    return m_keys[0] || m_keys[1] || m_keys[2] || m_keys[3];
+}
 
-bool Camera::moving() {
-    return m_keys.left || m_keys.right || m_keys.up || m_keys.down;
+void Camera::set_type(const CameraType type) {
+    m_type = type;
 }
 
-float Camera::near_clip() {
-    return m_z_near;
+CameraType Camera::get_type() const {
+    return m_type;
 }
 
-float Camera::far_clip() {
-    return m_z_far;
+void Camera::set_key_state(const CameraMovement key, const bool pressed) {
+    switch (key) {
+    case CameraMovement::FORWARD:
+        m_keys[0] = pressed;
+        break;
+    case CameraMovement::BACKWARD:
+        m_keys[2] = pressed;
+        break;
+    case CameraMovement::LEFT:
+        m_keys[1] = pressed;
+        break;
+    case CameraMovement::RIGHT:
+        m_keys[3] = pressed;
+        break;
+    }
```

Comment:
```suggestion
    m_keys[static_cast<std::underlying_type_t<int>>(key)] = pressed;
```

requireds `#include <type_traits>`

Reply:
I don't think `std::underlying_type` makes sense here.

Reply:
That doesn't work here.